src/traces/box/cross_trace_calc.js

```diff
@@ -90,23 +90,35 @@ function setPositionOffset(traceType, gd, boxList, posAxis, pad) {
     var groupgap = fullLayout[traceType + 'groupgap'];
     var padfactor = (1 - gap) * (1 - groupgap) * dPos / fullLayout[numKey];
 
+    // Find maximum trace width
+    // we baseline this at dPos
+    var max_half_width = dPos;
+    for(i = 0; i < boxList.length; i++) {
+        calcTrace = calcdata[boxList[i]];
+
+        if(calcTrace[0].trace.vwidth) {
+            if(calcTrace[0].trace.vwidth / 2 > max_half_width) {
+                max_half_width = calcTrace[0].trace.vwidth / 2;
+            }
+        }
+    }
+
     // autoscale the x axis - including space for points if they're off the side
     // TODO: this will overdo it if the outermost boxes don't have
     // their points as far out as the other boxes
     var extremes = Axes.findExtremes(posAxis, boxdv.vals, {
-        vpadminus: dPos + pad[0] * padfactor,
-        vpadplus: dPos + pad[1] * padfactor
+        vpadminus: max_half_width + pad[0] * padfactor,
+        vpadplus: max_half_width + pad[1] * padfactor
```

Comment:
Ah good catch - unfortunately the `TODO` concern above this gets more acute with custom widths. If you show two traces, one off to the left with a very narrow width and a wide one to the right, the one on the left will be padded as much as the one on the right, giving an inappropriately large autorange.

It *looks* to me as though we could pretty easily fix this by calculating `extremes` per trace. That would go below in the trace loop and might be as simple as:
```js
var thisDPos = calcTrace[0].t.dPos = (calcTrace[0].trace.width / 2) || dPos;
var positions = calcTrace.map(function(di) { return di.pos });
var extremes = Axes.findExtremes(posAxis, positions, {
    vpadminus: thisDPos + pad[0] * padfactor,
    vpadminus: thisDPos + pad[1] * padfactor
});
calcTrace[0].trace._extremes[posAxis._id] = extremes;
```
That wouldn't address the original `TODO` - which I guess is about a per-trace `pad` / `padfactor` depending on `trace.pointpos`, I'd have to look to see if there's an easy solution to that, there may be... but if you don't feel like digging into that feel free to just move the `TODO` along with this change.

Reply:
For

```javascript
  violin1 => x0=0.0, width=0.1
  violin2 => x0=0.1, width=1.0
```
![with changes](https://user-images.githubusercontent.com/10369095/47441046-5dd62800-d77d-11e8-9cf3-61dca34958b9.png)

and PRE your change above:

![before change](https://user-images.githubusercontent.com/10369095/47441674-a510e880-d77e-11e8-8db8-1ae26e7c16ff.png)

Reply:
so not that much improvement. We need a better algo for this.

Reply:
Ah, right - because we don't take `trace.side` into account - if it's `'positive'` we should set `vpadminus: 0`, if `'negative'` we should set `vpadplus: 0`.

Reply:
Muuuch better:

![newplot](https://user-images.githubusercontent.com/10369095/47455105-ff21a600-d79e-11e8-9931-2d24d144279b.png)